### DISCUSSION TITLE
RemoveVisitor Merges

### DISCUSSION THREAD

**Discussion by thomas-quadratic** [Quadratic-Labs/VerkleTries_Besu#37] [Ideas]
When using PutVisitor to put a value, we can split in the middle of an extension.
However, as currently implemented, RemoveVisitor does not do the inverse operation. That is, if we remove a leaf, we could be in a state where we have Nodes with single child. These could be merged into a single Node with an extension.

This way, if we put then remove, we should necessarily get the Trie we started with.
Although not essential, in general, it would be desirable that the structure of the Trie depends only on its key-value pairs and not on the history of operations performed on it.

If validated, this would lead to a new issue.

**Comment by matkt** [comment]
Indeed we have to find the tree as it was before. Otherwise the state will be invalid and bonsai will not be able to rollback . This is what is done on the PMT
